### PR TITLE
DRIVERS-2577: $readPreference is not sent to standalone server

### PR DESCRIPTION
**Pull request by jmikola**
https://jira.mongodb.org/browse/DRIVERS-2577

Please complete the following before merging:

- [x] Update changelog.
- [x] Make sure there are generated JSON files from the YAML test files.
- [x] Test changes in at least one language driver.
- [x] Test these changes against all server versions and topologies (including standalone, replica set, sharded clusters, and serverless).

POC PR in PHPLIB: https://github.com/mongodb/mongo-php-library/pull/1066

### CODE REVIEW

File: source/run-command/tests/unified/runCommand.yml [hunk 1]
```diff
@@ -87,6 +87,9 @@ tests:
               commandName: ping
 
   - description: attaches the provided $readPreference to given command
+    runOnRequirements:
+      # $readPreference is not sent to standalone servers
+      topologies: [ replicaset, sharded-replicaset, load-balanced, sharded ]
```
Comment (jmikola) [inline source/run-command/tests/unified/runCommand.yml:92]: [NODE-5207](https://jira.mongodb.org/browse/NODE-5207) was already resolved, so I'm not sure how this was missed unless you're either not testing on standalones or sending `$readPreference` to standalones. I believe the latter conflicts with the [Server Selection spec](https://github.com/mongodb/specifications/blob/master/source/server-selection/server-selection.rst#topology-type-single):

> Type Standalone: clients MUST NOT send the read preference to the server

I see the [Run Command spec](https://github.com/mongodb/specifications/blob/master/source/run-command/run-command.rst#readpreference) states:

> If the provided ReadPreference is NOT {mode: primary}, the command sent MUST include the $readPreference global command argument.

This sentence follows a reference to the Server Selection spec, but read on its own one could argue it conflicts. In contrast, [Global Command Arguments](https://github.com/mongodb/specifications/blob/master/source/message/OP_MSG.rst#global-command-arguments) in the OP_MSG spec defers to the Server Selection spec for rules on appending `$readPreference`.

**Review by alcaeus — APPROVED** [review]

**Review by jyemin — COMMENTED** [review]
Is there a test asserting that a "primary" read preference is not sent to any topology type?  If not, should we add one?

**Review by jyemin — COMMENTED** [review]
Is there a test asserting that a "primary" read preference is not sent to any topology type?  If not, should we add one?

**Review by jyemin — COMMENTED** [review]
Is there a test asserting that a "primary" read preference is not sent to any topology type?  If not, should we add one?

File: source/run-command/run-command.rst [hunk 2]
```diff
@@ -113,7 +113,7 @@ To facilitate server selection the RunCommand operation MUST accept an optional
 
 * See Server Selection's section on `Use of read preferences with commands <https://github.com/mongodb/specifications/blob/master/source/server-selection/server-selection.rst#use-of-read-preferences-with-commands>`_
 
-If the provided ReadPreference is NOT ``{mode: primary}``, the command sent MUST include the ``$readPreference`` global command argument.
+If the provided ReadPreference is NOT ``{mode: primary}`` and the selected server is NOT a standalone, the command sent MUST include the ``$readPreference`` global command argument.
```
[the remark this answers is not in the document]
Comment (nbbeeken) [reply source/run-command/run-command.rst:116]: I'm happy with it here, we still have the reference to server selection above

File: source/run-command/tests/unified/runCommand.yml [hunk 3]
```diff
@@ -105,6 +108,50 @@ tests:
...
+      - client: *client
+        events:
+          - commandStartedEvent:
+              command:
+                ping: 1
+                $readPreference: { $$exists: false }
+                $db: *db
+              commandName: ping
+
+  - description: does not attach primary $readPreference to given command
```
[the remark this answers is not in the document]
Comment (nbbeeken) [reply source/run-command/tests/unified/runCommand.yml:136]: The test `always attaches $db and implicit lsid to given command and omits default readPreference` covers this with the assumption that primary is always the default

**Review by nbbeeken — APPROVED** [review]

File: source/run-command/tests/unified/runCommand.yml (hunk 1 above)
Comment (nbbeeken) [reply source/run-command/tests/unified/runCommand.yml]: The Node driver is sending $readPreference to standalones

**Review by dariakp — APPROVED** [review]

File: source/run-command/tests/unified/runCommand.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (jmikola) [reply source/run-command/tests/unified/runCommand.yml:136]: Do you need this removed or is it worth the extra test coverage for an explicit `readPreference` option? I suppose it's possible for this to exercise a different code path in a driver.

[the remark this answers is not in the document]
Comment (nbbeeken) [reply source/run-command/tests/unified/runCommand.yml:136]: I think it's a good addition, I agree pulling in the default primary vs a caller passing it in is worth testing both are omitted from the command.

**Pull request merged**